New Resource: azurerm_data_factory_dataset_cosmosdb_mongoapi

gerrytan opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Adds a new resource for Data Factory CosmosDB MongoDB dataset. The resource is modelled similar to [data_factory_dataset_cosmosdb_sqlapi](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/data_factory_dataset_cosmosdb_sqlapi)

Resource relationship:

```
azurerm_data_factory --> azurerm_data_factory_linked_service_cosmosdb_mongoapi --> azurerm_data_factory_dataset_cosmosdb_mongoapi
```

REST API: https://learn.microsoft.com/en-us/rest/api/datafactory/datasets/create-or-update?view=rest-datafactory-2018-06-01

Resource screenshot in portal:


![image](https://github.com/user-attachments/assets/e89ef9ae-002f-4e37-bf00-8bcbf6a2a5a5)


## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

acctest log: [acctests_TestAccDataFactoryDatasetCosmosDbMongoDb_7358021176_20241205-1121+1100.log](https://microsoftapc-my.sharepoint.com/:u:/r/personal/tangerry_microsoft_com/Documents/Attachments/pr-attachments/azurerm-testing/acctests_TestAccDataFactoryDatasetCosmosDbMongoDb_7358021176_20241205-1121+1100.log?csf=1&web=1&e=riP3HX)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_data_factory_dataset_cosmosdb_mongoapi` - new resource [GH-27106]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Closes #27106 


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

gerrytan committed:
New resource for Data factory CosmosDB MongoDB API dataset

magodo commented on internal/services/datafactory/data_factory_dataset_cosmosdb_mongoapi_resource.go:
```diff
@@ -0,0 +1,339 @@
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+		"name": {
```
Can we reorder the schema here? Conventionally, the `name` and any parent resource name attributes come first, then those required ones (in alphabetic order), then the optional ones (alphabetically), then the computed ones.

gerrytan replied:
Fixed in 1179a4200a

magodo commented on internal/services/datafactory/data_factory_dataset_cosmosdb_mongoapi_resource.go:
```diff
@@ -0,0 +1,339 @@
+type DataFactoryDatasetCosmosDbMongoDbResource struct{}
+
+type DataFactoryDatasetCosmosDbMongoDbResourceModel struct {
+	AdditionalProperties map[string]interface{} `tfschema:"additional_properties"`
```
Probably we can change it to `map[string]string`?

gerrytan replied:
Fixed in 1179a4200a

magodo commented on internal/services/datafactory/data_factory_dataset_cosmosdb_mongoapi_resource.go:
```diff
@@ -0,0 +1,339 @@
+	Folder               string                 `tfschema:"folder"`
+	LinkedServiceName    string                 `tfschema:"linked_service_name"`
+	Name                 string                 `tfschema:"name"`
+	Parameters           map[string]interface{} `tfschema:"parameters"`
```
Probably we can change it to `map[string]string`?

gerrytan replied:
Fixed in 1179a4200a

magodo commented on internal/services/datafactory/data_factory.go:
```diff
@@ -97,6 +97,18 @@ func flattenDataFactoryParameters(input map[string]*datafactory.ParameterSpecifi
 	return output
 }
 
+func expandDataFactoryAnnotations(input *[]string) *[]interface{} {
```
You don't need this as there is one function called `utils.ExpandStringSlice` does the same thing.

gerrytan replied:
Nice! Fixed in 1179a4200a

magodo commented on internal/services/datafactory/data_factory_dataset_cosmosdb_mongoapi_resource_test.go:
```diff
@@ -0,0 +1,169 @@
+				},
+			},
+		},
+		data.ImportStep(),
```
Can we add one more step to update it back from complete to basic?

gerrytan replied:
Added in 1179a4200a

magodo reviewed (changes requested):
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

gerrytan committed:
Addressing PR feedbacks - iteration 1

- Reorder schema properties
- Change AdditionalProperties model type to map[string]string
- Change Parameters to map[string]string
- Use existing utils.FlattenStringSlice instead of creating a new function
- acctest: update: add 1 more step to update back to basic
- terrafmt fix for acctest and docs

gerrytan committed:
Merge branch 'main' into gerrytan/datafactory-cosmosdb-mongodb-new-resource

magodo commented:
LGTM! Thank you @gerrytan!

gerrytan commented:
Acctest passes as of commit f65b76f033 : [acctests_TestAccDataFactoryDatasetCosmosDbMongoDb_20241209-1628+1100.log](https://microsoftapc-my.sharepoint.com/:u:/r/personal/tangerry_microsoft_com/Documents/Attachments/pr-attachments/azurerm-testing/acctests_TestAccDataFactoryDatasetCosmosDbMongoDb_20241209-1628+1100.log?csf=1&web=1&e=bmsGbW)?csf=1&web=1&e=bmsGbW

magodo commented:
![{EA6F4C5D-DF40-43C0-9EB1-F1381BCD064C}](https://github.com/user-attachments/assets/9f3f165f-339e-4be9-a661-660477e00865)

jackofallops commented on internal/services/datafactory/data_factory_dataset_cosmosdb_mongoapi_resource.go:
```diff
@@ -0,0 +1,339 @@
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+	"github.com/jackofallops/kermit/sdk/datafactory/2018-06-01/datafactory"
```
Is there a reason we're using `kermit` here rather than `go-azure-sdk`? It appears supported. We should be avoiding any new usages of Kermit as we are attempting to deprecate and remove this workaround SDK.

gerrytan replied:
Have switched to go-azure-sdk in d95a18f654. There are tons of other resources in this folder still using kermit though, I've decided they're out of scope of this PR.

jackofallops commented on website/docs/r/data_factory_dataset_cosmosdb_mongoapi.html.markdown:
```diff
@@ -0,0 +1,87 @@
+
+---
+
+* `additional_properties` - (Optional) A map of additional properties to associate with the Data Factory Dataset.
```
As per `parameters`, can we make this a little more clear as to what is expected here, and what is supported?

gerrytan replied:
Have added more info to the doc in d95a18f654

jackofallops reviewed (changes requested):
Hi @gerrytan - Few questions and observations below, the key one is the choice of SDK for this new resource. We should prefer `go-azure-sdk` over `kermit` unless there's a compelling technical reason as we are attempting to discontinue the use of `kermit` entirely.
If you can take a look I'll continue review. Thanks!

gerrytan committed:
Addressing PR feedbacks - iteration 2

- Switch to go-azure-sdk
- Add nil checks, remove additional_properties from acctest
- Document the limited support of string parameters
- Renamed resource go type to be consistent with file name

gerrytan committed:
Merge branch 'main' into gerrytan/datafactory-cosmosdb-mongodb-new-resource

gerrytan commented on website/docs/r/data_factory_dataset_cosmosdb_mongoapi.html.markdown:
```diff
@@ -0,0 +1,87 @@
+
+* `folder` - (Optional) The folder that this Dataset is in. If not specified, the Dataset will appear at the root level.
+
+* `parameters` - (Optional) A map of parameters to associate with the Data Factory Dataset.
```
Have added additional info that only string -> string is supported. I decided not to support other parameter type to maintain consistency with the other `_dataset_` resources in the folder.

d95a18f654

gerrytan commented on internal/services/datafactory/data_factory_dataset_cosmosdb_mongoapi_resource.go:
```diff
@@ -0,0 +1,339 @@
+package datafactory
```
Good catch, have fixed this into `_mondodbapi_resource` and corresponding go model type.

d95a18f654

gerrytan committed:
Renamed variables to be more consistent with other methods

gerrytan commented:
acctest passed at the latest commit 645977dfbd: [acctests_TestAccDataFactoryDatasetCosmosDbMongoDbApi_645977dfbd_20241213-1238+1100.zip](https://microsoftapc-my.sharepoint.com/:u:/r/personal/tangerry_microsoft_com/Documents/Attachments/pr-attachments/azurerm-testing/acctests_TestAccDataFactoryDatasetCosmosDbMongoDbApi_645977dfbd_20241213-1238+1100.zip?csf=1&web=1&e=z18aUN)

gerrytan committed:
Merge branch 'main' into gerrytan/datafactory-cosmosdb-mongodb-new-resource

gerrytan commented:
Hello @jackofallops can you please give this PR from Dec 2024 a review again. All the changes you requested have been addressed, and I have merged latest main and rerun the acctest (output: [link](https://microsoftapc-my.sharepoint.com/:u:/g/personal/tangerry_microsoft_com/EYc8gBRDAhVEo2pmUVNFKKgBLOW_l7yv_qoD-qxZG2U64Q?email=ste%40hashicorp.com&e=Vnfw39))

gerrytan committed:
Merge branch 'main' into gerrytan/datafactory-cosmosdb-mongodb-new-resource

gerrytan committed:
Fix go vendoring

gerrytan committed:
Fix compile error

gerrytan committed:
Fix ordering

gerrytan commented:
Resolved conflict, rerun acctest: https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_DATAFACTORY/398023

gerrytan committed:
Merge branch 'main' into gerrytan/datafactory-cosmosdb-mongodb-new-resource
